ndk/src/midi.rs

```diff
@@ -0,0 +1,292 @@
+                .map_err(|e| MediaError::UnknownStatus(ffi::media_status_t(e.number as c_int)))?;
+            Ok(device_type)
+        } else {
+            Err(MediaError::from_status(ffi::media_status_t(device_type)).unwrap_err())
+        }
+    }
+
+    /// Opens the input port so that the client can send data to it.
+    pub fn open_input_port(&self, port_number: i32) -> Result<MidiInputPort> {
+        unsafe {
+            let input_port =
+                construct(|res| ffi::AMidiInputPort_open(self.ptr.as_ptr(), port_number, res))?;
+            Ok(MidiInputPort::from_ptr(NonNull::new_unchecked(input_port)))
+        }
+    }
+
+    /// Opens the output port so that the client can receive data from it.
+    pub fn open_output_port(&self, port_number: i32) -> Result<MidiOutputPort> {
+        unsafe {
+            let output_port =
+                construct(|res| ffi::AMidiOutputPort_open(self.ptr.as_ptr(), port_number, res))?;
+            Ok(MidiOutputPort::from_ptr(NonNull::new_unchecked(
+                output_port,
+            )))
+        }
+    }
+}
+
+impl Drop for MidiDevice {
+    fn drop(&mut self) {
+        let status = unsafe { ffi::AMidiDevice_release(self.ptr.as_ptr()) };
+        MediaError::from_status(status).unwrap();
+    }
+}
+
+pub struct MidiInputPort<'a> {
+    ptr: NonNull<ffi::AMidiInputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiInputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiInputPort")
+            .field("inner", &self.ptr)
+            .finish()
+    }
+}
+
+impl<'a> MidiInputPort<'a> {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::AMidiInputPort`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::AMidiInputPort>) -> Self {
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn ptr(&self) -> NonNull<ffi::AMidiInputPort> {
+        self.ptr
+    }
+
+    /// Sends data to this port.
+    pub fn send(&self, buffer: &[u8]) -> Result<usize> {
+        let num_bytes_sent =
+            unsafe { ffi::AMidiInputPort_send(self.ptr.as_ptr(), buffer.as_ptr(), buffer.len()) };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(MediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int)).unwrap_err())
+        }
+    }
+
+    /// Sends a message with a 'MIDI flush command code' to this port.
+    ///
+    /// This should cause a receiver to discard any pending MIDI data it may have accumulated and
+    /// not processed.
+    pub fn send_flush(&self) -> Result<()> {
+        let result = unsafe { ffi::AMidiInputPort_sendFlush(self.ptr.as_ptr()) };
+        MediaError::from_status(result)
+    }
+
+    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
```

Comment:
```suggestion
    /// Sends data to the this port with a timestamp.
    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
```

Do we also need extra docs what the timestamp is supposed to mean (and if it is e.g. monotonic with the system, or since the device was opened)? I have zero experience with MIDI so this may be common knowledge :)

Reply:
It's the return value of `long java.lang.System.nanoTime()`, but the NDK documents it as a `CLOCK_MONOTONIC` time... I initially tried to use `Instant`, but I'm not sure if I can assume that the current implementation of `Instant` on Unix systems will keep using `CLOCK_MONOTONIC`.

Reply:
Yeah don't cast into `Instant` unless it has a public API detailing how to manually create such a type. Please document those findings in a doc comment.

Sounds like `java.lang.System.nanoTime()` returns `CLOCK_MONOTONIC` (on Unix systems)? If there's a source I'd love to reference it in the doc-comment.

Reply:
Another question, just for my understanding: is this timestamp detailing when this message was put together (same for the `timestamp` field in `receive()`), or should it be some time in the future when the receiver should process/play back the given message/note?